cmd/ocsp-updater/main.go

```diff
@@ -107,96 +214,102 @@ func (updater *OCSPUpdater) processResponse(tx *gorp.Transaction, serial string)
+		if err != nil {
+			updater.log.AuditErr(fmt.Errorf("Failed to generate OCSP response: %s", err))
+			continue
+		}
+		responses = append(responses, meta)
 	}
 
+	tx, err := updater.dbMap.Begin()
+	if err != nil {
+		updater.log.AuditErr(err)
+		return
+	}
+	for _, meta := range responses {
+		err = updater.storeResponse(tx, meta)
+		if err != nil {
+			updater.log.AuditErr(fmt.Errorf("Failed to store OCSP response: %s", err))
+			tx.Rollback()
+			return
+		}
+	}
 	err = tx.Commit()
 	if err != nil {
-		updater.log.Err(fmt.Sprintf("OCSP %s: Error committing transaction, skipping: %s", serial, err))
-		updater.stats.Inc("OCSP.Updates.Failed", 1, 1.0)
-		tx.Rollback()
-		return err
+		updater.log.AuditErr(err)
+		return
 	}
 
-	updater.log.Info(fmt.Sprintf("OCSP %s: OK", serial))
-	updater.stats.Inc("OCSP.Updates.Processed", 1, 1.0)
-	updater.stats.TimingDuration("OCSP.Updates.UpdateLatency", time.Since(innerStart), 1.0)
-	return nil
+	return
 }
 
-// findStaleResponses opens a transaction and processes up to responseLimit
-// responses in a single batch. The responseLimit should be relatively small,
-// so as to limit the chance of the transaction failing due to concurrent
-// updates.
-func (updater *OCSPUpdater) findStaleResponses(oldestLastUpdatedTime time.Time, responseLimit int) error {
-	var certificateStatus []core.CertificateStatus
-	_, err := updater.dbMap.Select(&certificateStatus,
-		`SELECT cs.* FROM certificateStatus AS cs JOIN certificates AS cert ON cs.serial = cert.serial
-		 WHERE cs.ocspLastUpdated < ? AND cert.expires > now()
-		 ORDER BY cs.ocspLastUpdated ASC
-		 LIMIT ?`, oldestLastUpdatedTime, responseLimit)
-
-	if err == sql.ErrNoRows {
-		updater.log.Info("All up to date. No OCSP responses needed.")
-	} else if err != nil {
-		updater.log.Err(fmt.Sprintf("Error loading certificate status: %s", err))
-	} else {
-		updater.log.Info(fmt.Sprintf("Processing OCSP Responses...\n"))
-		outerStart := time.Now()
-
-		for i, status := range certificateStatus {
-			updater.log.Debug(fmt.Sprintf("OCSP %s: (%d/%d)", status.Serial, i, responseLimit))
-
-			err = updater.updateOneSerial(status.Serial)
-			// Abort if we recieve a fatal error
-			if _, ok := err.(FatalError); ok {
-				return err
-			}
-		}
+// oldOCSPResponsesTick looks for certificates with stale OCSP responses and
+// generates/stores new ones
+func (updater *OCSPUpdater) oldOCSPResponsesTick(batchSize int, now, _ time.Time) {
+	updater.rpcMu.RLock()
 
-		updater.stats.TimingDuration("OCSP.Updates.BatchLatency", time.Since(outerStart), 1.0)
-		updater.stats.Inc("OCSP.Updates.BatchesProcessed", 1, 1.0)
+	statuses, err := updater.findStaleOCSPResponses(now.Add(-updater.ocspMinTimeToExpiry), batchSize)
+	if err != nil && err != errNoStaleResponses {
+		updater.stats.Inc("OCSP.Updates.Failed", 1, 1.0)
+		updater.log.AuditErr(err)
+		updater.rpcMu.RUnlock()
+		return
 	}
 
-	return err
+	updater.generateOCSPResponses(statuses)
```

Comment:
Error stats here, maybe. would still be good to have context in the errors coming out of it.